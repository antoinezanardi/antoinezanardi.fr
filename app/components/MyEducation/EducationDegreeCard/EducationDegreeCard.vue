<template>
  <div class="card">
    <div class="row">
      <div
        class="bg-primary col-md-3"
        data-aos="fade-right"
        data-aos-duration="500"
        data-aos-offset="50"
      >
        <div class="card-body cc-education-header">
          <PeriodTimeline
            class="education-timeline"
            does-show-year-only
            :finished-at="educationDegree.degree.obtainedAt"
            :image="schoolImage"
            :image-alt="schoolImageAlt"
            :period-dates-aria-label="periodDatesAriaLabel"
            :started-at="educationDegree.degree.startedAt"
            :url="educationDegree.school.url"
          />
        </div>
      </div>

      <div
        class="col-md-9"
        data-aos="fade-left"
        data-aos-duration="500"
        data-aos-offset="50"
      >
        <div class="card-body">
          <h3
            class="degree-name font-weight-bold h4 mb-2 mt-0"
          >
            {{ educationDegree.degree.name }}
          </h3>

          <div class="d-flex">
            <p
              class="category font-weight-bold mb-0 school-name"
            >
              {{ schoolLabel }}
            </p>

            <CountryFlag
              class="ms-2 school-flag"
              :country="educationDegree.school.country"
            />
          </div>

          <hr>

          <p
            v-for="(paragraph, index) in educationDegree.degree.description"
            :key="index"
            class="degree-description"
          >
            {{ paragraph }}
          </p>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import type { EducationDegreeCardProps } from "~/components/MyEducation/EducationDegreeCard/education-degree-card.types";
import CountryFlag from "~/components/shared/Images/CountryFlag/CountryFlag.vue";
import PeriodTimeline from "~/components/shared/Period/PeriodTimeline.vue";

const props = defineProps<EducationDegreeCardProps>();

const { t } = useI18n();

const schoolImage = computed<string>(() => props.educationDegree.school.image ?? "college-icon.png");

const schoolLabel = computed<string>(() => {
  const { school } = props.educationDegree;
  let label = school.translatedName ?? "?";
  if (school.city !== undefined) {
    label += `, ${school.city}`;
  }
  return label;
});

const schoolImageAlt = computed<string>(() => props.educationDegree.school.translatedName ?? "?");

const periodDatesAriaLabel = computed<string>(() => {
  const finishedAtYear = props.educationDegree.degree.obtainedAt.getFullYear().toString();

  return t("EducationDegreeCard.degreeObtainedAtAriaLabel", {
    degreeName: props.educationDegree.degree.name,
    obtainedAt: finishedAtYear,
  });
});
</script>

<style lang="scss" scoped>
.cc-education-header {
  padding-top: 25px !important;
}
@media (max-width: 768px) {
  .cc-education-header {
    padding-right: 1.25rem !important;
  }
}
</style>