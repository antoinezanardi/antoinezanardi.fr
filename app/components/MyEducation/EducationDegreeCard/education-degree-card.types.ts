import type { EducationDegree } from "~/models/education-degree/education-degree.types";

type EducationDegreeCardProps = {
  educationDegree: EducationDegree;
};

export type { EducationDegreeCardProps };