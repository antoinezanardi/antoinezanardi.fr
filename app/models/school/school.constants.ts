import type { School, SchoolName } from "~/models/school/school.types";

const SCHOOL_NAMES = [
  "vimeu",
  "epitech",
  "laval",
  "cloudNativeComputingFoundation",
  "google",
] as const;

const SCHOOLS: Record<SchoolName, School> = Object.freeze({
  vimeu: {
    name: "vimeu",
    city: "Friville-Escarbotin",
    country: "france",
    url: "https://lycee-vimeu-friville.ac-amiens.fr/",
  },
  epitech: {
    name: "epitech",
    city: "Lille",
    country: "france",
    url: "https://www.epitech.eu/",
    image: "epitech-logo.png",
  },
  laval: {
    name: "laval",
    city: "Qu√©bec",
    country: "canada",
    url: "https://www.ulaval.ca/",
    image: "laval-logo.png",
  },
  cloudNativeComputingFoundation: {
    name: "cloudNativeComputingFoundation",
    country: "usa",
    url: "https://www.cncf.io/",
    image: "cloud-native-computed-foundation-logo.png",
  },
  google: {
    name: "google",
    country: "usa",
    url: "https://cloud.google.com/",
    image: "google-cloud-platform-logo.png",
  },
});

export { SCHOOL_NAMES, SCHOOLS };