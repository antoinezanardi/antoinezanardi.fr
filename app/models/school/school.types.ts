import type { TupleToUnion } from "type-fest";

import type { Country } from "~/models/country/country.types";
import type { SCHOOL_NAMES } from "~/models/school/school.constants";

type SchoolName = TupleToUnion<typeof SCHOOL_NAMES>;

type School = {
  name: SchoolName;
  translatedName?: string;
  city?: string;
  country: Country;
  image?: string;
  url: string;
};

export type { School, SchoolName };