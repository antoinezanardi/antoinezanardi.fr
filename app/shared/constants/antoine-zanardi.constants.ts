const ANTOINE_ZANARDI_FULL_NAME = "Antoine ZANARDI";

const ANTOINE_ZANARDI_BIRTH_DATE = "1996-04-14";

const ANTOINE_ZANARDI_GITHUB_URL = "https://github.com/antoinezanardi/";

const ANTOINE_ZANARDI_GITHUB_AVATAR_URL = "https://avatars.githubusercontent.com/u/11405981?v=4";

const ANTOINE_ZANARDI_LINKEDIN_URL = "https://www.linkedin.com/in/antoinezanardi/";

export {
  ANTOINE_ZANARDI_FULL_NAME,
  ANTOINE_ZANARDI_BIRTH_DATE,
  ANTOINE_ZANARDI_GITHUB_URL,
  ANTOINE_ZANARDI_GITHUB_AVATAR_URL,
  ANTOINE_ZANARDI_LINKEDIN_URL,
};