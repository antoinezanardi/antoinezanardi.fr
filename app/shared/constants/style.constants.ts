const PRIMARY_COLOR = "#12671D";

const BLACK_COLOR = "#000000";

const WHITE_COLOR = "#FFFFFF";

export {
  PRIMARY_COLOR,
  BLACK_COLOR,
  WHITE_COLOR,
};