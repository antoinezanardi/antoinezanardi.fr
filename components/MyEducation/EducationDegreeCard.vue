<template>
  <div class="card">
    <div class="row">
      <div class="col-md-3 bg-primary" data-aos="fade-right" data-aos-offset="50" data-aos-duration="500">
        <div class="card-body cc-education-header">
          <PeriodTimeline
            :image="schoolImage"
            :url="educationDegree.school.url"
            :started-at="educationDegree.degree.startedAt"
            :finished-at="educationDegree.degree.obtainedAt"
            does-show-year-only
          />
        </div>
      </div>
      <div class="col-md-9" data-aos="fade-left" data-aos-offset="50" data-aos-duration="500">
        <div class="card-body">
          <div class="h4" v-html="educationDegree.degree.name"/>
          <div class="d-flex">
            <p class="category mb-0" v-html="schoolLabel"/>
            <CountryFlag class="ms-2" :country="educationDegree.school.country"/>
          </div>
          <hr/>
          <p v-for="(paragraph, index) of educationDegree.degree.description" :key="index" v-html="paragraph"/>
        </div>
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { computed } from "#imports";
import CountryFlag from "~/components/shared/CountryFlag.vue";
import PeriodTimeline from "~/components/shared/Period/PeriodTimeline.vue";
import type { EducationDegree } from "~/models/EducationDegree";

const props = defineProps<{
  educationDegree: EducationDegree;
}>();

const schoolImage = computed<string>(() => props.educationDegree.school.image ?? "college-icon.png");
const schoolLabel = computed<string>(() => {
  const { school } = props.educationDegree;
  let label = school.translatedName ?? "?";
  if (school.city !== undefined) {
    label += `, ${school.city}`;
  }
  return label;
});
</script>

<style lang="scss" scoped>
.cc-education-header {
  padding-top: 25px !important;
}
@media (max-width: 768px) {
  .cc-education-header {
    padding-right: 1.25rem !important;
  }
}
</style>