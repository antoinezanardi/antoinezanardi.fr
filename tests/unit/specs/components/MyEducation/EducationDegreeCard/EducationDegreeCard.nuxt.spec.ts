import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import type PeriodTimeline from "~/components/shared/Period/PeriodTimeline.vue";
import type { EducationDegreeCardProps } from "~/components/MyEducation/EducationDegreeCard/education-degree-card.types";
import EducationDegreeCard from "~/components/MyEducation/EducationDegreeCard/EducationDegreeCard.vue";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type CountryFlag from "~/components/shared/Images/CountryFlag/CountryFlag.vue";
import { SCHOOLS } from "~/models/school/school.constants";
import type { EducationDegree } from "~/models/education-degree/education-degree.types";

describe("Education Degree Card Component", () => {
  const defaultProps: EducationDegreeCardProps = {
    educationDegree: {
      degree: {
        name: `Degrees.ITMasterDegree`,
        description: ["MyEducation.firstThreeYearsInEpitech", "MyEducation.lastYearsInEpitech"],
        startedAt: new Date("2014-01-01T00:00:00.000Z"),
        obtainedAt: new Date("2019-01-01T00:00:00.000Z"),
      },
      school: {
        ...SCHOOLS.epitech,
        translatedName: `Schools.${SCHOOLS.epitech.name}`,
      },
    },
  };
  let wrapper: ReturnType<typeof mount<typeof EducationDegreeCard>>;

  async function mountEducationDegreeCardComponent(options: ComponentMountingOptions<typeof EducationDegreeCard> = {}):
  Promise<ReturnType<typeof mount<typeof EducationDegreeCard>>> {
    return mountSuspendedComponent(EducationDegreeCard, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountEducationDegreeCardComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Period Timeline", () => {
    it("should pass education degree started date when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".education-timeline");

      expect(periodTimeline.props("startedAt")).toBe(defaultProps.educationDegree.degree.startedAt);
    });

    it("should pass education degree obtained date when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".education-timeline");

      expect(periodTimeline.props("finishedAt")).toBe(defaultProps.educationDegree.degree.obtainedAt);
    });

    it("should pass school image when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".education-timeline");

      expect(periodTimeline.props("image")).toBe(defaultProps.educationDegree.school.image);
    });

    it("should pass school image alt as translated name when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".education-timeline");

      expect(periodTimeline.props("imageAlt")).toBe(defaultProps.educationDegree.school.translatedName);
    });

    it("should pass interrogative image alt when school translated name is not found.", async() => {
      const educationDegreeWithoutTranslatedName: EducationDegree = {
        ...defaultProps.educationDegree,
        school: {
          ...defaultProps.educationDegree.school,
          translatedName: undefined,
        },
      };
      wrapper = await mountEducationDegreeCardComponent({ props: { educationDegree: educationDegreeWithoutTranslatedName } });
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".education-timeline");

      expect(periodTimeline.props("imageAlt")).toBe("?");
    });

    it("should pass default school image when school image is not set in education degree.", async() => {
      const educationDegreeWithoutImage: EducationDegree = {
        ...defaultProps.educationDegree,
        school: {
          ...SCHOOLS.epitech,
          image: undefined,
        },
      };
      wrapper = await mountEducationDegreeCardComponent({ props: { educationDegree: educationDegreeWithoutImage } });
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".education-timeline");

      expect(periodTimeline.props("image")).toBe("college-icon.png");
    });

    it("should pass school url when rendered.", () => {
      const periodTimeline = wrapper.findComponent<typeof PeriodTimeline>(".education-timeline");

      expect(periodTimeline.props("url")).toBe(defaultProps.educationDegree.school.url);
    });
  });

  describe("Degree", () => {
    it("should render degree name when rendered.", () => {
      const degreeName = wrapper.find<HTMLDivElement>(".degree-name");

      expect(degreeName.text()).toBe("Degrees.ITMasterDegree");
    });
  });

  describe("School", () => {
    it("should render school name with city when city is set in school props.", () => {
      const schoolLabel = wrapper.find<HTMLDivElement>(".school-name");

      expect(schoolLabel.text()).toBe("Schools.epitech, Lille");
    });

    it("should render school name without city when city is not set in school props.", async() => {
      const educationDegreeWithoutCity: EducationDegree = {
        ...defaultProps.educationDegree,
        school: {
          ...defaultProps.educationDegree.school,
          city: undefined,
        },
      };
      wrapper = await mountEducationDegreeCardComponent({ props: { educationDegree: educationDegreeWithoutCity } });
      const schoolLabel = wrapper.find<HTMLDivElement>(".school-name");

      expect(schoolLabel.text()).toBe("Schools.epitech");
    });

    it("should render interrogative when school translated name is not found.", async() => {
      const educationDegreeWithoutTranslatedName: EducationDegree = {
        ...defaultProps.educationDegree,
        school: {
          ...defaultProps.educationDegree.school,
          translatedName: undefined,
          city: undefined,
        },
      };
      wrapper = await mountEducationDegreeCardComponent({ props: { educationDegree: educationDegreeWithoutTranslatedName } });
      const schoolLabel = wrapper.find<HTMLDivElement>(".school-name");

      expect(schoolLabel.text()).toBe("?");
    });

    it("should pass school country to country flag when rendered.", () => {
      const countryFlag = wrapper.findComponent<typeof CountryFlag>(".school-flag");

      expect(countryFlag.props("country")).toBe(defaultProps.educationDegree.school.country);
    });
  });

  describe("Paragraphs", () => {
    it("should render degree paragraphs when rendered.", () => {
      const degreeParagraphs = wrapper.findAll<HTMLParagraphElement>(".degree-description");

      expect(degreeParagraphs).toHaveLength(2);
      expect(degreeParagraphs[0].text()).toBe("MyEducation.firstThreeYearsInEpitech");
      expect(degreeParagraphs[1].text()).toBe("MyEducation.lastYearsInEpitech");
    });
  });
});