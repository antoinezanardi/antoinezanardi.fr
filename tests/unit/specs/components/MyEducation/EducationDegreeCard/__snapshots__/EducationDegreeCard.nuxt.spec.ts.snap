// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Education Degree Card Component > should match snapshot when rendered. 1`] = `
"<div data-v-91841a15="" class="card">
  <div data-v-91841a15="" class="row">
    <div data-v-91841a15="" class="bg-primary col-md-3" data-aos="fade-right" data-aos-duration="500" data-aos-offset="50">
      <div data-v-91841a15="" class="card-body cc-education-header">
        <period-timeline-stub data-v-91841a15="" image="epitech-logo.png" imagealt="Schools.epitech, Lille" url="https://www.epitech.eu/" doesshowyearonly="true" startedat="Wed Jan 01 2014 00:00:00 GMT+0000 (Coordinated Universal Time)" finishedat="Tue Jan 01 2019 00:00:00 GMT+0000 (Coordinated Universal Time)" class="education-timeline"></period-timeline-stub>
      </div>
    </div>
    <div data-v-91841a15="" class="col-md-9" data-aos="fade-left" data-aos-duration="500" data-aos-offset="50">
      <div data-v-91841a15="" class="card-body">
        <h4 data-v-91841a15="" class="degree-name h4">Degrees.ITMasterDegree</h4>
        <div data-v-91841a15="" class="d-flex">
          <p data-v-91841a15="" class="category mb-0 school-name">Schools.epitech, Lille</p>
          <country-flag-stub data-v-91841a15="" country="france" class="ms-2 school-flag"></country-flag-stub>
        </div>
        <hr data-v-91841a15="">
        <p data-v-91841a15="" class="degree-description">MyEducation.firstThreeYearsInEpitech</p>
        <p data-v-91841a15="" class="degree-description">MyEducation.lastYearsInEpitech</p>
      </div>
    </div>
  </div>
</div>"
`;
